actors/builtin/miner/miner_state.go

```diff
@@ -957,6 +985,133 @@ func (st *State) checkPrecommitExpiry(store adt.Store, sectors abi.BitField) (de
 	return depositToBurn, nil
 }
 
+type AdvanceDeadlineResult struct {
+	PledgeDelta                                abi.TokenAmount
+	PowerDelta                                 PowerPair
+	UndeclaredFaultyPower, DeclaredFaultyPower PowerPair
+}
+
+// AdvanceDeadline advances the deadline. It:
+// - Processes expired sectors.
+// - Handles missed proofs.
+// - Returns the changes to power & pledge, and faulty power (both declared and undeclared).
+func (st *State) AdvanceDeadline(store adt.Store, currEpoch abi.ChainEpoch) (*AdvanceDeadlineResult, error) {
+	pledgeDelta := abi.NewTokenAmount(0)
+	powerDelta := NewPowerPairZero()
+
+	undeclaredFaultyPower := NewPowerPairZero()
+	declaredFaultyPower := NewPowerPairZero()
+
+	// Note: because the cron actor is not invoked on epochs with empty tipsets, the current epoch is not necessarily
+	// exactly the final epoch of the deadline; it may be slightly later (i.e. in the subsequent deadline/period).
+	// Further, this method is invoked once *before* the first proving period starts, after the actor is first
+	// constructed; this is detected by !dlInfo.PeriodStarted().
+	// Use dlInfo.PeriodEnd() rather than rt.CurrEpoch unless certain of the desired semantics.
+
+	dlInfo := st.DeadlineInfo(currEpoch)
+	// Period not started? Nothing to do.
+	if !dlInfo.PeriodStarted() {
+		return &AdvanceDeadlineResult{
+			pledgeDelta,
+			powerDelta,
+			undeclaredFaultyPower,
+			declaredFaultyPower,
+		}, nil
+	}
+
+	// Advance to the next deadline (in case we short-circuit below).
+	st.CurrentDeadline = (st.CurrentDeadline + 1) % WPoStPeriodDeadlines
+	if st.CurrentDeadline == 0 {
+		st.ProvingPeriodStart = st.ProvingPeriodStart + WPoStProvingPeriod
+	}
+
+	deadlines, err := st.LoadDeadlines(store)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load deadlines: %w", err)
+	}
+	deadline, err := deadlines.LoadDeadline(store, dlInfo.Index)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load deadline %d: %w", dlInfo.Index, err)
+	}
+
+	// No live sectors in this deadline, nothing to do.
+	if deadline.LiveSectors == 0 {
+		return &AdvanceDeadlineResult{
+			pledgeDelta,
+			powerDelta,
+			undeclaredFaultyPower,
+			declaredFaultyPower,
+		}, nil
+	}
+
+	quant := dlInfo.QuantSpec()
+	{
+		// Detect and penalize missing proofs.
+		faultExpiration := dlInfo.Last() + FaultMaxAge
+
+		powerDelta, undeclaredFaultyPower, err = deadline.ProcessDeadlineEnd(store, quant, faultExpiration)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to process end of deadline %d: %w", dlInfo.Index, err)
+		}
+		declaredFaultyPower = deadline.FaultyPower.Sub(undeclaredFaultyPower)
```

Comment:
I don't understand this, or at least it's appearance here. No "declared faults" can happen while processing missed PoSts. In combination with my confusion above, this line would make a bit more sense at the bottom of this method. But if it's just trying to count the faulty power before the new faults were added .... just read `deadline.FaultyPower` before processing the new faults?